feat: add common properties with different nullability to base mixin

DevNico opened the pull request:
The new field `isCommonWithDifferentNullability` might not be perfect but it was the best way I could come up with to propagate the information.

Close #715

DevNico commented:
I just also had the thought that we could also expose non-common properties that don't have a type collision as nullable on the base mixin. What do you think? @rrousselGit

DevNico commented:
In theory, we could also add all properties that are not common in all constructors and do not have a type conflict to the base mixin.

What do you think? @rrousselGit?

DevNico commented:
Also instead of removing properties with different nullability, we could add them but only allow non-null inputs.

DevNico commented:
> Also instead of removing properties with different nullability, we could add them but only allow non-null inputs.

I went ahead and implemented that feature as well

TimWhiting commented:
I don't think this closes #715 which is requesting for a common base class, whereas this just addresses the nullability aspect. This is a great first step though!

DevNico commented:
You're right I read that issue a while ago and needed this specific feature now so I went ahead and implemented it. I vaguely remembered reading something about it and therefore just quickly added the closing statement without re-reading the original issue.

Removed the closing statement since it is out of scope for what I currently require and would also require a lot more work.

rrousselGit commented:
Hello! Thanks for your work!


Supporting cases like:

```dart
factory Foo.first(int a) = _First;
factory Foo.second(double a) = _First;
...


Foo foo;
num a = foo.a;
```

should not be as complex as I originally thought.


I later learned about `TypeSystem`/`TypeProvider` (which can be found on `LibraryElement` I believe). They provide utilities find the nearest common type between two properties

It's probably worth using that directly instead.

DevNico commented:
If you haven't started working on that I can give it a go and try to refactor this PR to use `TypeSystem`/`TypeProvider`

rrousselGit commented:
Sure, go ahead 😄

DevNico commented:
Properties with a common super type that isn't `Object` now have getters in the base mixin. 🎉

I have not included them in the `copyWith` since we can only upcast to the common supertype but not downcast.

DevNico commented:
I also noticed that the `super` constructors could be refactored to make use of the `super.` shorthand. Is that something you'd be alright with? I could do that in this PR or open another one for those changes.

rrousselGit commented:
Great!

The super syntax isn't used yet because Freezed supports an sdk version lower than 2.17

It's not very important considering the code is generated

rrousselGit commented:
I'm a bit busy until the Flutter Viking talks the 1st of September. Not sure I'll be able to properly review this PR before

But thanks a lot for your work 😊

TimWhiting commented on packages/freezed/lib/src/freezed_generator.dart:
```diff
@@ -248,26 +240,48 @@ Read here: https://github.com/rrousselGit/freezed/blob/master/packages/freezed/C
-            return parameter.copyWith(isFinal: true);
+            final commonTypeWithMatch =
+                parameter.parameterElement!.library!.typeSystem.leastUpperBound(
+              parameter.parameterElement!.type,
```
Should this be commonType? (Initialize commonType with parameter.parameterElement!.type). That way it can handle >2 constructors with a common base type

TimWhiting replied:
You could then remove the if (commonType != commonTypeWithMatch) check below.

DevNico replied:
We would keep getting the supertype with each previous supertype until we hit Object. And if Object isn't the initial type we'll return since as per my other statement I think it's not desirable to have each common parameter as Object in the base mixin.

TimWhiting commented on packages/freezed/lib/src/freezed_generator.dart:
```diff
@@ -248,26 +240,48 @@ Read here: https://github.com/rrousselGit/freezed/blob/master/packages/freezed/C
+              matchingParameter.parameterElement!.type,
+            );
+
+            if (commonTypeWithMatch.isDartCoreObject) return null;
```
Why return early here? Object is more specific than Object? (it gives you a non-null object, but I guess there isn't much you can do with Object). I would assume that if someone named all the parameters the same there could be a case made for wanting access to that parameter even if it returns Object?

DevNico replied:
That is correct I think there should be an additional check to see if the original type was `Object`. The early return is there because Object is the supertype of everything so all "incompatible" types would be Object in the base mixin which in my opinion wouldn't be desirable.

TimWhiting commented on packages/freezed/lib/src/freezed_generator.dart:
```diff
@@ -248,26 +240,48 @@ Read here: https://github.com/rrousselGit/freezed/blob/master/packages/freezed/C
+
+            if (!allowCommonSuperType &&
+                commonTypeWithMatch.getDisplayString(withNullability: false) !=
+                    parameter.type?.replaceAll(r'?$', '')) return null;
```
Why not `parameter.parameterElement.type.getDisplayString(withNullability: false)`, instead of manually removing nullability? Additionally shouldn't the `!allowCommonSuperType` check occur for determining the `commonType` as well? Specifically in the copyWith below here you might want to only use `commonType` if `allowCommonSuperType` is true.

TimWhiting commented on packages/freezed/test/integration/single_class_constructor.dart:
```diff
@@ -450,7 +450,7 @@ class NullDefault with _$NullDefault {
 class ExplicitConstDefault with _$ExplicitConstDefault {
   factory ExplicitConstDefault(
       //ignore: unnecessary_const
-      [@Default(const <Object>[]) List<Object> value]) = _ExplicitConstDefault;
```
Why was this removed? As you can see with the lint, and the name of this class, this is testing support for the explicit const keyword here.

DevNico replied:
Good catch that was the autoformatter removing the const.

TimWhiting commented on packages/freezed/test/multiple_constructors_test.dart:
```diff
@@ -504,8 +504,65 @@ void main() {
+void main() {
+  final param = SharedParamCommonSuperType('a', 'b', 42);
+  param.copyWith(b: 42);
+}
```
This probably could be done, (not a common super type, but by finding a common sub type).

For example in the null-case:
String?
String
common subtype == String, which is why copyWith a String makes sense because it works for all variations of the parameter.

For a not much more complicated hierarchy.
num
double
common subtype == double -> you can use double for the parameter

Finally, in general:
if (param is subtype of otherConstructorParam):
  commonSubtype = param
else if (otherConstructorParam is subtype of param):
  commonSubtype = otherConstructorParam
else
  commonSubtype = Never (no common subtype)


Additionally this test fails to compile not because of the copyWith but because of the String passed into the constructor I think

DevNico replied:
The test was a bad rewrite of the earlier test for the different nullability. Yes it fails due to the incorrect string constructor paramter but it would also fail for the parameter which it was intended for.

I'll fix the test and also try to find a way of generally finding common subtypes.

TimWhiting commented:
@rrousselGit I'm happy to help review PRs if you are busy with other things. Excited for this PR!

rrousselGit commented:
@TimWhiting Sure go ahead!

I should be a bit more available by September.

DevNico commented:
@rrousselGit all two types have `Object` / `Object?` as a common supertype (except for `dynamic` maybe? I'd have to check that) would you say it is desirable to have those parameters in the common base mixin as `Object`?

As an example take:
```dart
factory Foo.first(String a) = _First;
factory Foo.second(double a) = _Second;

Foo foo;
Object a = foo.a;
```

rrousselGit commented:
If all the cases possess the same variable, I think showing Object/Object? is fine.
No dynamic though, that's for sure.

miDeb commented on packages/freezed/lib/src/freezed_generator.dart:
```diff
@@ -248,26 +240,48 @@ Read here: https://github.com/rrousselGit/freezed/blob/master/packages/freezed/C
 
-          if (hasAnyFinalProperty) {
-            return parameter.copyWith(isFinal: true);
+            final commonTypeWithMatch =
```
So, if I'm reading this correctly, this code repeatedly checks the least upper bound for the parameter in the first constructor and the parameter in another constructor. Only if this always returns the same result it will be considered.

I think this could be improved by declaring a `commonType` outside of the `for` loop, initializing it to the first parameter type, and then in the loop for each parameter type updating the `commonType` to the least upper bound between the `commonType` so far and the current parameter type.

That way the least upper bound between all parameter types will be found, even if that is different from the least upper bound between the first constructor and the second. Consider this example:

```dart
class A {}
class B implements A {}
class C implements B {}
class D implements A {}

@freezed
class Container {
  const factory Container.one({required B param}) = _One;
  const factory Container.two({required C param}) = _Two;
  const factory Container.three({required D param}) = _Three;
}
```
The types `B`, `C` and `D` are all subclasses of `A`, but currently this is not recognized.

DevNico replied:
Tim already pointed that out. I have added a test for your use case and the parameter is available as expected.

miDeb replied:
Sorry, I missed that.

miDeb commented:
We should probably still do a comparison by the string representation of the type, to make sure this doesn't break:
```dart
@freezed
class Container with _$Container {
  const factory Container.one() = _One;
  const factory Container.two() = _Two;
  const factory Container.three() = _Three;
}

@freezed
class ContainerWrapper with _$ContainerWrapper {
  const factory ContainerWrapper.one({required _One a}) = _WOne;
  const factory ContainerWrapper.two({required _One a}) = _WTwo;
}
```

`One` is not recognized because it is not yet generated, but we should generate a common getter as it is currently the case.

miDeb commented:
There is also a bug when handling typedefs: Consider this example (to be placed into the integration test directory)
```dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'external_typedef.dart';
import 'external_typedef_two.dart' as two;

@freezed
class ContainsTypedefs with _$ContainsTypedefs {
  const factory ContainsTypedefs.one(
      {required ExternalTypedef a, required two.ExternalTypedefTwo b}) = _TOne;
  const factory ContainsTypedefs.two(
      {required ExternalTypedef? a, required two.ExternalTypedefTwo? b}) = _TTwo;
}
```

The compiled code has compile errors because we apparently "forget" that this is a typedef after searching for the least upper bound.

DevNico commented:
I'll fix those and any other cases you might find and add the respective tests later

miDeb commented:
Here's another convoluted one for you :wink: 
```dart
@freezed
class CommonType with _$CommonType {
  const factory CommonType.one({required int a}) = _One;
  const factory CommonType.two({required num a}) = _Two;
}

@freezed
class CommonTypeContainer with _$CommonTypeContainer {
  const factory CommonTypeContainer({required CommonType internal}) = _CommonTypeContainer;
}

void main() {
  final container = CommonTypeContainer(
    internal: CommonType.two(a: 0.0),
  );
  container.copyWith.internal();
}
```

This fails at runtime due to the type cast inserted [here](https://github.com/rrousselGit/freezed/pull/740/files#diff-d617954163786415754a92b9e87692ac0567ac0b028230ea42bb3246231370d4R208). If you remove it, the compiler will tell you that something is wrong: Because the common `copyWith` is generated using a common subtype, calling it with anything other than this common subtype (or _its_ subtypes) is not allowed. However, some union variants might contain a value that is not this common subtype, and if you don't specify an argument for you `copyWith` call this value will be used as the default value, which will then fail.

rrousselGit commented:
CopyWith should not include properties that have a type difference

DevNico commented:
> CopyWith should not include properties that have a type difference

So no common subtypes? But would you still want types of different nullability?

TimWhiting commented:
> CopyWith should not include properties that have a type difference

Why? 
For example in the null-case:
String?
String
common subtype == String, which is why copyWith passing a String makes sense because it works for all variations of the parameter. (The original PR had this)

For a not much more complicated hierarchy.
num
double
common subtype == double -> you can use double for the parameter, because no matter which constructor / union was used, double makes sense.

miDeb commented:
> Unlike a common supertype(leastUpperBound) which is used for getters, common subtypes don't end up with something general like Object?

Exactly. For common subtypes it will only consider the types that the constructors use, so if one constructor uses a type that is a subtype of all the types the other constructors use, this type can be used for the general copyWith implementation.

miDeb commented:
`unfreezed` classes don't work yet:

```dart
@unfreezed
class CommonType with _$CommonType {
  const factory CommonType.one({required int a}) = _One;
  const factory CommonType.two({required num a}) = _Two;
}
```
produces:

The return type of getter 'a' is 'num' which isn't a subtype of the type 'int' of its setter 'a'.
Try changing the types so that they are compatible.dart [getter_not_subtype_setter_types](https://dart.dev/diagnostics/getter_not_subtype_setter_types)

DevNico commented:
> ```dart
> @freezed
> class CommonType with _$CommonType {
>   const factory CommonType.one({required int a}) = _One;
>   const factory CommonType.two({required num a}) = _Two;
> }
> ```

That's actually quite tricky. I feel like the only solution would be to use the integrated map function and perform the `copyWith` on each concrete class. Not sure if there are any other ways of doing this.

Instead of the not compiling / running
```dart
return _then(_value.copyWith(
  a: a == freezed
      ? _value.a
      : a // ignore: cast_nullable_to_non_nullable
          as int,
));
```

We'd have something like
```dart
return _then(_value.map(
  one: (value) => value.copyWith(
    a: a == freezed
        ? value.a
        : a // ignore: cast_nullable_to_non_nullable
            as int,
  ),
  two: (value) => value.copyWith(
    a: a == freezed
        ? value.a
        : a // ignore: cast_nullable_to_non_nullable
            as num,
  ),
));
```

Do you guys have any other suggestions and or opinions on this? @TimWhiting @miDeb @rrousselGit

DevNico commented:
I fixed all the other cases except for the `copyWith`. For that one, I'll wait for you to give your opinion.

TimWhiting commented:
@DevNico 
Turns out that it has the same error for null/non-null case as well. CopyWith is very hard it seems. I think that your alternative makes sense, but I think it would probably be good to not generate the `.map` unless there is any parameter with a common subtype. Additionally, I think that this feature would be good to independently enable / disable. (Maybe call it enhancedCopyWith)?

Additionally, I keep wondering why you changed some of the lists to Iterables? Can you explain the reason for that change?

DevNico commented:
I changed them to Iterable since it was just unnecessary in some places to do the toList since the Iterable is sufficient. AFAIK it's a little bit more efficient not to do the toList when not needed

DevNico commented:
I'd really like to wait for feedback from @rrousselGit here. I do agree with you @TimWhiting that an `enhancedCopWith` would make sense. We can add all common supertypes to the base mixin by default as a new feature.

The common subtype `enhancedCopyWith` using `map` would then only be generated when this new opt-in flag is turned on _and_ a property with a common subtype exists.

There is also the option of just always generating the `map` call when there is a common subtype e.g. not having the feature be opt-in. This might have performance implications tho which would have to be evaluated whether these are negligible or not.

rrousselGit commented:
I don't really have time to give feedback at the moment.

But I think copyWith support would better be done in a separate PR. 
copyWith on different properties would likely involve a lot of changes to get it right. And it has a bunch of edge-cases 

Like:

```dart
@freezed
class CommonType with _$CommonType {
  const factory CommonType.one({required int a}) = _One;
  const factory CommonType.two({required num a}) = _Two;
}
```

vs:

```dart
@freezed
class CommonType with _$CommonType {
  const factory CommonType.one({required int a}) = _One;
  const factory CommonType.two({required double a}) = _Two;
}
```

or:

```dart
@freezed
class A with _$A {
  const factory A(int count) = _A;
}

@freezed
class B with _$B {
  const factory B(int count) = _B;
}

@freezed
class CommonType with _$CommonType {
  const factory CommonType.one(A value) = _One;
  const factory CommonType.two(B value) = _Two;
}

void main() {
  CommonType type;

  type = type.copyWith.value(count: 42);
}
```

rrousselGit commented on packages/freezed/lib/src/templates/abstract_template.dart:
```diff
@@ -13,15 +13,17 @@ class Abstract {
   final Data data;
   final CopyWith? copyWith;
-  final List<Property> commonProperties;
+  final Iterable<Property> commonProperties;
```
Same for the other places where this was changed

DevNico replied:
I changed every output / variable back to List.

rrousselGit commented on packages/freezed/lib/src/templates/copy_with.dart:
```diff
@@ -65,23 +63,26 @@ ${_abstractDeepCopyMethods().join()}
-            );
-          }).toList(),
+          namedParameters: allProperties
+              .map((e) => Parameter(
```
Don't use => for big function blocks like that.

{ return } is voluntary

rrousselGit commented on packages/freezed/lib/src/templates/properties.dart:
```diff
@@ -106,13 +136,16 @@ class Property {
-        doc: doc,
-        body: ' => throw $privConstUsedErrorVarName;',
-      );
+  Setter? get unimplementedSetter =>
```
I don't like that this is nullable. It doesn't seem like the correct place to put logic here.

rrousselGit commented on packages/freezed/lib/src/utils.dart:
```diff
@@ -71,3 +71,16 @@ String _fixCase(String input, String separator) =>
       return lower;
     });
+
+String typeStringWithoutNullability(String typeString) {
```
Why do we need those utilities? `getDisplayString` has a `withNullability` parameter

DevNico replied:
That is true but for not generated types (e.g. other Freezed classes) the common type checking is done on Strings not DartTypes.

rrousselGit commented on packages/freezed/test/common_types_test.dart:
```diff
@@ -0,0 +1,231 @@
+    test(
+      'copy has shared params with common subtypes available',
+      () async {
+        await expectLater(compile(r'''
```
For new tests checking compilation errors, it'd be great to use my expect_error package instead

This  `compile`  is technical dept and should be removed at some point.

rrousselGit commented on packages/freezed/test/common_types_test.dart:
```diff
@@ -0,0 +1,231 @@
+  });
+
+  group('CommonTypedefs', () {
+    test('generates correct typedefs', () async {
```
What is this test trying to do?

In general I don't like the idea of analyzing the source code like that. Chances are there's a better way to do this that doesn't rely on the super complex analyzer API

DevNico replied:
I literally just copied the other typedef test and adapted it here. If you don't like it like that I can rewrite it to be like the other tests

miDeb commented:
> Do you guys have any other suggestions

I could come up with another solution (not sure if we should continue this discussion in this PR though):

We could amend the concrete implementations of `freezed` classes by a new private function that allows to supply the result type and the `then` function to its `CopyWith` class. Something like this:
```dart
  @override
  @JsonKey(ignore: true)
  _$$_OneCopyWith<$Res> _$copyWith<$Res>($Res Function(_$_One) _then) =>
      __$$_OneCopyWithImpl<$Res>(this, _then);
```

Then, instead of generating a common concrete `copyWith` implementation the abstract class could return the specific concrete `CopyWith` class:

```dart
  factory $CommonTypeCopyWith(
          CommonType value, $Res Function(CommonType) then) =>
      value._$copyWith<$Res>(then);
```

DevNico commented:
I removed variables where the common subtype doesn't match the common supertype e.g. it won't work without refactoring copyWith so this pr can be merged as just the common getters feature.

The improved version of the common copyWith can then be implemented in a later PR.

TimWhiting commented on packages/freezed/lib/src/freezed_generator.dart:
```diff
@@ -438,8 +482,8 @@ Read here: https://github.com/rrousselGit/freezed/blob/master/packages/freezed/C
   Iterable<CloneableProperty> _commonCloneableProperties(
-    List<ConstructorDetails> constructors,
-    List<Property> commonProperties,
+    Iterable<ConstructorDetails> constructors,
```
There are a couple of Iterable changes still in this PR

DevNico replied:
I left these in because when used as a parameter there is no lazy iterable implication but giving more flexibility in the inputs to the function.

TimWhiting commented on packages/freezed/lib/src/freezed_generator.dart:
```diff
@@ -247,27 +229,89 @@ Read here: https://github.com/rrousselGit/freezed/blob/master/packages/freezed/C
+                commonSupertypeString =
+                    typeStringWithNullability(commonSupertypeString);
+                commonSubtypeString =
+                    typeStringWithoutNullability(commonSupertypeString);
```
should this be `typeStringWithoutNullability(commonSubtypeString);`

DevNico replied:
I'll take a look at this again next week!

TimWhiting commented on packages/freezed/lib/src/templates/copy_with.dart:
```diff
@@ -197,10 +200,17 @@ $s''';
+
+      final type = p.commonSubtype ?? p.type;
+      if (p.commonSubtype != null && p.type != p.commonSubtype) {
+        ternary += 'as $type ';
```
I could be wrong, but isn't this still doing the invalid casts for copyWith for a common subtype of `num, int -> int`?

Are we moving the copyWith changes to another PR or are you still trying to do the `T & T? -> T` subtyping (which doesn't run into cast errors right?)

DevNico replied:
This is only for the nullability subtyping which shouldn't run into problems at all.

rrousselGit commented on packages/freezed/test/common_types_test.dart:
```diff
@@ -0,0 +1,173 @@
+      expect(value.c, 56.78);
+    });
+
+    // test('copy has all with common subtypes available', () {
+    //   var value = const CommonSuperSubtype('a', 1337, 42);
+    //   expect(value.a, 'a');
+
+    //   value = value.copyWith(a: 'b', c: 24.42);
+    //   expect(value.a, 'b');
+    //   expect(value.c, 24.42);
+    // });
+
+    // test('copy has shared params with common subtypes available', () async {
+    //   var param = const CommonSuperSubtype('a', 1337, 42);
+    //   param = param.copyWith(a: '2', c: 24.42);
+    //   expect(param.a, '2');
+    //   expect(param.c, 24.42);
+    // });
+
+//     test('copy does not have shared params without common subtype available',
+//         () async {
+//       await expectLater(library.withCode('''
+// import 'integration/common_types.dart';
+
+// void main() {
+//   final param = CommonSuperSubtype('a', 1337, 42);
+
+//   // expect-error: UNDEFINED_NAMED_PARAMETER
+//   param.copyWith(b: 42);
+// }
+// '''), compiles);
+//     });
```
```suggestion
```

rrousselGit commented on packages/freezed/test/common_types_test.dart:
```diff
@@ -0,0 +1,173 @@
+
+import 'integration/common_types.dart';
+
+// TODO: Uncomment tests when copyWith for common subtypes is implemented
```
```suggestion
```

rrousselGit commented on packages/freezed/test/common_types_test.dart:
```diff
@@ -0,0 +1,143 @@
+    });
+  });
+
+  // group('CommonTypeNestedContainer', () {
+  //   test(
+  //     'nested copyWith with a default value that is not a subtype of the common subtype',
+  //     () {
+  //       var commonTypeContainer = const CommonTypeNestedContainer(
+  //         internal: CommonTypeNested.two(a: 0.0),
+  //       );
+  //       commonTypeContainer = commonTypeContainer.copyWith.internal();
+  //       expect(commonTypeContainer.internal.a, 0.0);
+
+  //       commonTypeContainer = commonTypeContainer.copyWith.internal(a: 12);
+  //       expect(commonTypeContainer.internal.a, 12);
+  //     },
+  //   );
+  // });
+
```
```suggestion
```

rrousselGit commented on packages/freezed/test/common_types_test.dart:
```diff
@@ -0,0 +1,122 @@
+  group('CommonSuperSubtype', () {
+    test('has the common properties available', () {
+      var value = const CommonSuperSubtype('a', 1337, 42.24);
+      expect(value.a, 'a');
```
This should test that the public type of those properties is respectively `String?`, `num` and `num`

Technically this test could expose those properties as `Object?`, and the test would still pass

rrousselGit commented on packages/freezed/test/common_types_test.dart:
```diff
@@ -0,0 +1,122 @@
+
+  group('CommonUnfreezed', () {
+    test('should have common getters available', () {
+      var value = CommonUnfreezed.one(a: 42, b: 3.14);
```
Same as the other test. We want to make sure that shared getting are correctly typed.

rrousselGit commented:
There are a few conflicts and some small tests changes. But otherwise LGTM!

Thanks for your work 😄

rrousselGit commented:
Looking back at the logic for obtaining the common subtype between union cases, the logic seems to do more than necessary

Typedefs and generated types are not supported by Freezed in general. Trying to support it for this specific case doesn't make sense, as they are still unsupported for everything else.

rrousselGit commented:
Alright, time to merge this!

Sorry for the delay :)
